### PR TITLE
spmatmul sparse matrix multiplication - performance improvements

### PR DESCRIPTION
**Pull request by KlausC**
The implementation of Gustavson's algorithm used to need a sorting-step ofter the result matrix is constructed. The algorithm has been modified to produce the result vectors properly ordered.
Speed improvement is about 4 times at comparable space usage.
old results:
```
julia> n = 10000;
julia> Random.seed!(0); A = sprandn(n, n, 0.01); b = randn(n);  nnz(A)
1000598
julia> @benchmark A * A
BenchmarkTools.Trial: 
  memory estimate:  977.61 MiB
  allocs estimate:  28
  --------------
  minimum time:     5.815 s (3.76% GC)
  median time:      5.815 s (3.76% GC)
  mean time:        5.815 s (3.76% GC)
  maximum time:     5.815 s (3.76% GC)
  --------------
  samples:          1
  evals/sample:     1
```
new results:
```
julia> @benchmark A * A
BenchmarkTools.Trial: 
  memory estimate:  1.04 GiB
  allocs estimate:  10
  --------------
  minimum time:     1.299 s (3.74% GC)
  median time:      1.328 s (5.37% GC)
  mean time:        1.327 s (5.55% GC)
  maximum time:     1.353 s (7.64% GC)
  --------------
  samples:          4
  evals/sample:     1
```

### CODE REVIEW

**Comment by jebej** [comment]
this doesn't seem to work with `SparseVector`s

**Review by ViralBShah — CHANGES_REQUESTED** [review]

**Comment by KristofferC** [comment]
I had some benchmarks in https://github.com/JuliaLang/julia/pull/29682 that could be used here as well.

File: stdlib/SparseArrays/src/linalg.jl [hunk 1]
```diff
@@ -147,61 +147,81 @@ end
...
                 end
             end
-            for vp in colptrC[i]:(ip - 1)
-                nzvalC[vp] = x[rowvalC[vp]]
+            for k in 1:mA
+                if xb[k]
+                    nzvalC[ip] = x[k]
+                    rowvalC[ip] = k
+                    ip += 1
+                end
```
[the remark this answers is not in the document]
Comment (KlausC) [reply stdlib/SparseArrays/src/linalg.jl]: I see your point. My standard example was not representative for really big and really sparse matrices. I will try to combine the best of both.

**Comment by ViralBShah** [comment]
Just a thought - we can consider having an optional parameter for accumulating into a dense vector or by sorting. The default would have to be the one that is asymptotically good, but this dense accumulator approach could be picked optionally.

**Comment by KlausC** [comment]
Hi @ViralBShah,
I adapted your first proposal and found a robust and simple way to make the decision between full scan (my first proposal in this PR for the less sparse corner cases) and the sorting of the row indices (the original software which handles the asymptotic case better - huge matrix with strong sparsity).
The decision is made for each result column separately.
Interestinly the decision depends on `nnz * log2(nnz) * 3 < m`, where `m` is the result column length and `nnz` is the number of nonzeros in one column.

Some additional changes are:
- initial estimation of nnz of result
- replace Vector{Int} by Vector{Bool} to tag nonzero slots.
- re-use nzvals as dense akkumulator storage.
- use in-place sort! for each result column individually - removed `sparseMatrixSort!` at end.
As a result all observed cases so far have speed accelerations.

[evaluation plot](https://github.com/KlausC/SparseWrappers.jl/blob/master/data/sort_or_scan0_28.png)
This picture shows the evaluation of the alternatives. x-axis is `log2(m)`, y-axis is `log2(nnz)`. The blue points have better performance for the scanning approach, the yellow point for the sorting approach. The solid line is the graph of nnz ->  3 nnz*log2(nnz).

Benchmark results:
The measurements were done in a copy of the original software `SparseWrappers.spmatmul_orig` and the modified version `SparseWrappers.spmatmul`. The last one is identical to the version in this PR.

My own original example:
```
julia> n = 10000;
julia> Random.seed!(0); A = sprandn(n, n, 0.01); b = randn(n);  nnz(A)
1000598
julia> @btime SparseWrappers.spmatmul_orig($A, $A);
  5.259 s (28 allocations: 122.60 MiB)
julia> @btime SparseWrappers.spmatmul($A, $A);
  1.400 s (8 allocations: 1.04 GiB)
```

The asymptotic case of @ViralBShah 
```
julia> n = 10^6
1000000
julia> Random.seed!(0); A = sprandn(n, n, 10 / n); b = randn(n);  nnz(A)
10003563
julia> @btime SparseWrappers.spmatmul_orig($A, $A);
  9.613 s (24 allocations: 2.43 GiB)
julia> @btime SparseWrappers.spmatmul($A, $A);
  6.449 s (9 allocations: 1.65 GiB)

julia> Random.seed!(0); A = sprandn(n, n, 1 / n); b = randn(n);  nnz(A)
1000757
julia> @btime SparseWrappers.spmatmul_orig($A, $A);
  220.634 ms (20 allocations: 106.86 MiB)
julia> @btime SparseWrappers.spmatmul($A, $A);
  162.402 ms (11 allocations: 42.20 MiB) 
```

The benchmarks proposed by @KristofferC 
```
julia> for N = (^).(10, 1:7)
           S = spdiagm(0 => ones(N), 1 => -ones(N-1));
           SA = sparse(S')
           @show N
           @btime SparseWrappers.spmatmul_orig($S, $SA)
           @btime SparseWrappers.spmatmul($S, $SA)
       end
N = 10
  782.991 ns (10 allocations: 1.80 KiB)
  674.351 ns (7 allocations: 2.22 KiB)
N = 100
  6.028 μs (10 allocations: 11.81 KiB)
  4.024 μs (5 allocations: 8.11 KiB)
N = 1000
  55.681 μs (12 allocations: 110.41 KiB)
  36.536 μs (7 allocations: 77.89 KiB)
N = 10000
  544.767 μs (18 allocations: 1.07 MiB)
  356.850 μs (8 allocations: 775.78 KiB)
N = 100000
  5.501 ms (18 allocations: 10.68 MiB)
  3.685 ms (9 allocations: 7.57 MiB)
N = 1000000
  56.884 ms (18 allocations: 106.81 MiB)
  54.472 ms (9 allocations: 75.72 MiB)
N = 10000000
  995.334 ms (18 allocations: 1.04 GiB)
  620.350 ms (9 allocations: 757.22 MiB)

julia> using MatrixDepot # v0.8.0
populating internal database...

julia> md = mdopen("Wissgott/parabolic_fem")
(RS Wissgott/parabolic_fem(#1853)  525825x525825(3674625/2100225) 2007 [A, b] 'computational fluid dynamics problem' [Parabolic FEM, diffusion-convection reaction, constant homogenous diffusion]()

julia> A = sparse(md.A); # MatrixDepot delivers a Symmetric{SparseMatrixCSC}
julia> @btime SparseWrappers.spmatmul_orig($A, $A);
  422.052 ms (20 allocations: 248.35 MiB)
julia> @btime SparseWrappers.spmatmul($A, $A);
  273.490 ms (9 allocations: 435.51 MiB)
```

**Comment by ViralBShah** [comment]
This is some really nice work @KlausC!

**Review by ViralBShah — APPROVED** [review]

File: stdlib/SparseArrays/test/sparse.jl [hunk 2]
```diff
@@ -318,8 +318,7 @@ end
         a = sprand(10, 5, 0.7)
         b = sprand(5, 15, 0.3)
         @test maximum(abs.(a*b - Array(a)*Array(b))) < 100*eps()
-        @test maximum(abs.(SparseArrays.spmatmul(a,b,sortindices=:sortcols) - Array(a)*Array(b))) < 100*eps()
-        @test maximum(abs.(SparseArrays.spmatmul(a,b,sortindices=:doubletranspose) - Array(a)*Array(b))) < 100*eps()
+        @test maximum(abs.(SparseArrays.spmatmul(a,b) - Array(a)*Array(b))) < 100*eps()
```
Comment (ViralBShah) [inline stdlib/SparseArrays/test/sparse.jl]: This test is redundant with the one above, I believe. We should delete and maybe add another that tests the transpose multiplication case.

**Review by ViralBShah — CHANGES_REQUESTED** [review]

File: stdlib/SparseArrays/test/sparse.jl (hunk 2 above)
Comment (KlausC) [reply stdlib/SparseArrays/test/sparse.jl]: I will work on that tomorrow. I did not look at the `*` function yet. What do you mean exactly by the transpose case? `A * B'` or `A' * B` or both (if A and B are SparseMatrixCSC).

Comment (ViralBShah) [reply stdlib/SparseArrays/test/sparse.jl]: Yes, I believe that `a*b` and `spmatmul(a,b)` are exactly the same. The `*` just calls `spmatmul`, and hence it is the same test:

https://github.com/JuliaLang/julia/blob/855959124dbafadafdcd324771ec06831bc79a1a/stdlib/SparseArrays/src/linalg.jl#L142

But good to test other code paths for `a*b'` and `a'*b` with something like:
```
@test maximum(abs.(a'*b - Array(a')*Array(b))) < 100*eps()
@test maximum(abs.(a*b' - Array(a)*Array(b'))) < 100*eps()
```

**Comment by ViralBShah** [comment]
@KlausC The tests should be an easy fix. I also propose that we mention in the NEWS file about this performance enhancement, and that we no longer have the two options for `spmatmul`. Since `spmatmul` is in an `stdlib` and also not an exported function, I think we should be ok with this.

**Comment by ViralBShah** [comment]
Also, for future reference, here's a paper by @aydinbuluc on state of the art on sparse matrix multiplication: https://people.eecs.berkeley.edu/~aydin/spgemm_p2s218.pdf

**Comment by ViralBShah** [comment]
The `sortSparseMatrixCSC!` has this logic in it to pick different sorting algorithms based on the nnz in the column. Would it be possible to try it out in your performance tests and see if it helps?

```
        if numrows <= 16
            alg = Base.Sort.InsertionSort
        else
            alg = Base.Sort.QuickSort
        end
```

**Comment by KlausC** [comment]
> ... performance tests and see if it helps?
> 
> ```
>         if numrows <= 16
>             alg = Base.Sort.InsertionSort
>         else
>             alg = Base.Sort.QuickSort
>         end
> ```

I tried it and did not see a difference. The QuickSort algorithm itself branches to InsertionSort, if numrows <= 20, so the best we could gain is one `invoke`.

**Comment by ViralBShah** [comment]
Thanks!

**Comment by ViralBShah** [comment]
I am actually happy to merge this as is (the test changes are not that important).

**Review by ViralBShah — APPROVED** [review]

**Comment by KlausC** [comment]
I committed added test cases and NEWS.md after your merged. How do I get these last changes saved?

**Pull request merged**